Automattic/vip-go-mu-plugins
# Filter `woocommerce_prevent_admin_access` to not block valid users from wp-admin

## Description

To determine if `wp-admin` access should be blocked, WooCommerce is checking the capabilities on the user and not the user's actual role. Since the VIP 2FA plugin alters the user's capabilities so early, when WooCommerce checks them to see if the user should be able to access `wp-admin`, valid users who should have `wp-admin` access are being blocked and redirected to the front end of the site.

I don't think it's unreasonable for WooCommerce to assume that a user's capabilities _should_ match the capabilites of their role. We can use the `woocommerce_prevent_admin_access` filter to introduce a strict check to see that a user's _role_ has the capabilities that WooCommerce checks for and determine whether or not the user should have access to `wp-admin`.

This commit adds that filter, returns early if the VIP 2FA plugin is not enabled OR if WooCommerce has already determined that the user should have `wp-admin` access. It also introduces a helper function that takes an array of capabilites, looks through all roles for that capability, and returns an array of role slugs for _any_ role that contains _any_ of the passed capabilities. If the current user's role matches any role in the returned array, we grant that user access to `wp-admin`.

Fixes #1413 

## Checklist

Please make sure the items below have been covered before requesting a review:


## Steps to Test

1. On a VIP Go Environment, create a user with the `administrator` role without setting up 2FA.
1. Install and activate WooCommerce.
1. Try to login with the `administrator` user.
1. Verify the user can access `wp-admin`.

To make sure this doesn't break existing WooCommerce logic for users that should _always_ be blocked from `wp-admin`:

1. Create a user with the `subscriber` role on the above VIP Go environment.
1. Try to login with the `subscriber` role.
1. Verify that the user is redirected away from `wp-admin`.
